### PR TITLE
coredump: fix bug that loses core dump files when core dumps are compressed and disk space is low.

### PR DESCRIPTION
**Pull request by hgwalles**
Previously the save_external_coredump function returned a file
descriptor corresponding to the dumped file.  This descriptor was used
for two different purposes by calling code: a) access to the raw core
dump data; b) testing candidate files (via inode comparisons) while
vacuuming to protect the current core dump against vacuuming.

The descriptor returned always corresponded to a file containing the raw
core dump data.  However if compression was used and the core dump was
compressed then the descriptor returned did not correspond to the file
that would eventually be left on disk (ie the compressed file).  Thus
the file was never protected by vacuuming.  When disk space was scarce
all core dumps including the current one would be vacuumed and the
corresponding log message referred to a file that no longer existed.

This resulted in the following error message from coredumpctl if the
missing core dump was requested:
   Cannot retrieve coredump from journal nor disk.
   Failed to retrieve core: No such file or directory

With this patch, save_external_coredump now returns two descriptors, one to be used for
raw data access, and one to be used for inode comparisons to prevent
overzealous vacuuming.  Corresponding usage of these returned
descriptors also updated.

### CODE REVIEW

File: src/journal/coredump.c [hunk 1]
```diff
@@ -795,7 +807,7 @@ int main(int argc, char* argv[]) {
         coredump_vacuum(-1, arg_keep_free, arg_max_use);
 
         /* Always stream the coredump to disk, if that's possible */
-        r = save_external_coredump(info, uid, &filename, &coredump_fd, &coredump_size);
+        r = save_external_coredump(info, uid, &filename,&coredump_node_fd, &coredump_fd, &coredump_size);
```
Comment (zonque) [inline src/journal/coredump.c]: Please place a whitespace before each `&`.

**Comment by hgwalles** [comment]
Checks shouldn't have failed.  Restart checks?

**Comment by poettering** [comment]
Hmm, the closing confused semaphore... 

Could you please merge the whitespace fix into the main commit? i.e. rebase and merge the two commits with "git rebase -i"...

**Comment by poettering** [comment]
And then force-push, which should hopefully restart the semaphore logic.

**Comment by hgwalles** [comment]
I did consider both the alternatives (duping the file descriptor for the uncompressed case, or returning an invalid desciptor as a sentinel in the uncompressed case).
- I rejected the 'invalid descriptor' approach because, as pointed out above, it complicates the logic in the calling code.
- I rejected the 'dup the descriptor' approach because, in the uncompressed case there would now be two descriptors floating around for the same open file which is not what callers (ie, me!) usually expect.  Though it's not a problem now I can imagine callers somehow interleaving operations on the descriptors and getting confused by the results.

That said, if I'm just being paranoid and a dup seems the better choice, I'd be quite happy to replace the open with a fcntl(fd, FD_DUP_CLOEXEC, 3) (and possibly a warning comment).  I would still argue against returning an invalid descriptor in *ret_data_fd in the uncompressed case since that unnecessarily complicates the caller's logic.

Will merge the two commits shortly.  Thanks for taking a look.

**Comment by poettering** [comment]
fds are expensive and scarce, I'd really prefer if you'd return an invalidated fd here, and check for this in the caller. Not because it would really matter here, but mostly out of a matter of principle...

**Comment by hgwalles** [comment]
If minimizing resources used is the overriding principle then the invalidated fd approach would seem best.  I'll happily revise the patch accordingly.

**Comment by hgwalles** [comment]
Revised fix does not use excess fds, and is makes fewer textual changes, too.

**Comment by poettering** [comment]
Could you please rebase this and merge the two patches into one?

**Comment by hgwalles** [comment]
Sorry.  Done.

**Pull request merged**